src/dataflow/src/source/postgres.rs

```diff
@@ -188,71 +211,83 @@ impl PostgresSourceReader {
                         // publication, so we ignore them
                         Insert(event) if event.rel_id() != table_info.rel_id => continue,
                         Update(event) if event.rel_id() != table_info.rel_id => continue,
                         Delete(event) if event.rel_id() != table_info.rel_id => continue,
                         Insert(insert) => {
-                            let tx = tx.as_mut().ok_or_else(|| {
-                                anyhow!("got row event without a prior BEGIN event")
-                            })?;
-                            let row = self.row_from_tuple(insert.tuple())?;
-                            tx.insert(row).await?;
+                            let row = try_fatal!(self.row_from_tuple(insert.tuple()));
+                            inserts.push(row);
                         }
                         Update(update) => {
-                            let tx = tx.as_mut().ok_or_else(|| {
-                                anyhow!("got row event without a prior BEGIN event")
-                            })?;
-                            let old_row = self.row_from_tuple(update.old_tuple().unwrap())?;
-                            tx.delete(old_row).await?;
-
-                            let new_row = self.row_from_tuple(update.new_tuple())?;
-                            tx.insert(new_row).await?;
+                            let old_tuple = try_fatal!(update
+                                .old_tuple()
+                                .ok_or_else(|| anyhow!("full row missisng from update")));
+                            let old_row = try_fatal!(self.row_from_tuple(old_tuple));
+                            deletes.push(old_row);
+
+                            let new_row = try_fatal!(self.row_from_tuple(update.new_tuple()));
+                            inserts.push(new_row);
                         }
                         Delete(delete) => {
-                            let tx = tx.as_mut().ok_or_else(|| {
-                                anyhow!("got row event without a prior BEGIN event")
-                            })?;
-                            let row = self.row_from_tuple(delete.old_tuple().unwrap())?;
-                            tx.delete(row).await?;
+                            let old_tuple = try_fatal!(delete
+                                .old_tuple()
+                                .ok_or_else(|| anyhow!("full row missisng from delete")));
+                            let row = try_fatal!(self.row_from_tuple(old_tuple));
+                            deletes.push(row);
                         }
                         Commit(commit) => {
-                            self.lsn = commit.commit_lsn().into();
-                            // Release the lease
-                            tx = None;
+                            self.lsn = commit.end_lsn().into();
+
+                            let tx = timestamper.start_tx().await;
```

Comment:
This changed! Would you mind walking me through why the right thing to do is the deletes and then the inserts? Also, why we no longer need the lease released? Happy to do this over video if it's long to write up.

Reply:
The order of deletes vs inserts is not important since they are all timestamped the same. I just thought maybe timely/differential behave better if we first issue retractions? No idea. But the semantics are identical.

Regarding the lease, in the previous version of the code we were emitting the rows into the dataflow as they arrived from the replication stream. So we grabbed the lease when we saw a BEGIN, then used it for all subsequent messages, and then released the lease on COMMIT. The current version doesn't emit anything in the dataflow until a complete transaction is buffered in memory. So essentially the lease is taken and released within the COMMIT match branch.

The rationale for this change is that previously it could happen that the replication connection drops in the middle of receiving a transaction. But when we reconnect and restart the replication we'll receive the messages starting at the BEGIN of the last seen transaction, so we'd see duplicate messages and we'd have to avoid re-issuing them in the dataflow.

At first I thought maybe I can keep around an offset for the current in-flight transaction (something like "we're at the 4th changed row in the tx"), but I haven't verified that the messages within a transaction are always going to arrive at the same order. So I opted for this simpler solution where we first make sure we have a complete transaction in memory and only then issue it in the dataflow.

I hope this makes sense but if not happy to discuss it over video

Reply:
Totally makes sense, thank you!